### PR TITLE
[WIP] Add Sockeye baselines

### PR DESCRIPTION
**Pull request by Proyag**
*Issue #, if available:*

*Description of changes:*
- Cleaned up environment creation
  - Use `environment.yml` for conda setup; `conda create --file requirements.txt` does not work for pip packages
  - Removed `requirements.txt`
  - Added repos like Prism and subword-nmt as submodules instead of manually cloning
- Cleaned up data extraction
  - Directly extract json files into the directory we want instead of using `extract_all_jsons.py` to move them.
  - Removed `extract_all_jsons.py`
  - Added BPE files for dataset creation
- Added target factor generation scripts
  - Refactored `build_dataset.py` to use argparse, and extra options, like to generate segment duration info in a separate file (needed for factors)
  - Added `separate_factors.py` to generate target factor files
- Added Sockeye baselines
  - Model with interleaved phonemes and durations
  - Model with target factors, clean data
  - Model with target factors, noised data
  - Scripts and instructions to decode and evaluate these baselines
  - Scripts to reproduce these baselines

*TODO:*
- Add dub generation scripts for Sockeye output
- Change Sockeye links to the main repo pending https://github.com/awslabs/sockeye/pull/1082

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -19,26 +19,35 @@ limitations under the License.
...
+conda env create --file environment.yml
 conda activate iwslt-autodub
 
 cd <TODO_repo_dir>
+# TODO: We already pip install subword-nmt and should not need this
 git clone https://github.com/rsennrich/subword-nmt.git subword-nmt
 git clone https://github.com/moses-smt/mosesdecoder.git mosesdecoder
 
+# Download CoVoST2 TSV files
 aws s3 cp <TODO_dir1>/covost-2/covost_v2.en_de.train.tsv ./covost_tsv
```
Comment (mtresearcher) [inline README.md]: lets take this out and ask users to download covost data in a directory and use the TSV files extracted in that directory.

File: build_datasets.py [hunk 2]
```diff
@@ -1,247 +1,289 @@
...
+        "Run get_durations_frequencies.py first to get the dictionary of durations" \
+        " and how many times each is observed in our data!"
+    with open(args.durations_path, 'rb') as f:
+        logging.info("Loading durations' frequencies")
+        durations_pkl = pickle.load(f)
+    if not os.path.exists(args.processed_output_dir):
+        os.makedirs(args.processed_output_dir)
+
+    output_path = os.path.join(args.processed_output_dir, args.de_output_type)
+    if args.num_bins == 0:
```
Comment (mtresearcher) [inline build_datasets.py:260]: Is there a case where we do not bin at all? I would add a warning here that Binning is turned off.

File: separate_factors.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    Context Manager to neatly manage all the factor output file handles
+    """
+    def __init__(self, prefix):
+        self.handles = dict()
+        self.prefix = prefix
+
+    def __enter__(self):
+        for f in FACTOR_TYPES:
+            output_path = self.prefix + f.lower()
+            self.handles[f] = open(output_path, 'w')
```
Comment (mtresearcher) [inline separate_factors.py:48]: can you check if file exists already? and if it does either add a warning before opening in write mode or skip writing to it at all (assuming the file is correct).

File: separate_factors.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+            logging.info(f"Writing {f} factors to {output_path}")
+        self.handles['text'] = open(self.prefix + 'text', 'w')
```
Comment (mtresearcher) [inline separate_factors.py]: same here, either add warning or skip writing.

**Review by mtresearcher — COMMENTED** [review]
Code quality looks good. I left a few minor comments, will review in depth as we work through this.

File: separate_factors.py (hunk 3 above)
Comment (Proyag) [reply separate_factors.py:48]: Skipping individual files is trickier, but changed it to raise `FileExistsError` if file exists already, with a `--allow-overwrite` option to explicitly allow overwriting

File: build_datasets.py (hunk 2 above)
Comment (Proyag) [reply build_datasets.py:260]: I included the option when I thought we might use it, but didn't actually end up using it. Added a warning as you suggest.

File: README.md (hunk 1 above)
Comment (Proyag) [reply README.md]: Changed instructions to download and extract TSV files as in https://github.com/facebookresearch/covost#covost-2

**Pull request merged**